packages/react-apollo-hooks/src/useQuery.ts

```diff
@@ -0,0 +1,202 @@
+
+import {AsyncQueryComponentType} from '@shopify/react-graphql';
+
+import {
+  QueryHookOptions,
+  QueryHookResult,
+  QueryDocumentResolving,
+} from './types';
+import {useApolloClient} from './ApolloContext';
+import useGraphQLDocument from './useGraphQLDocument';
+import {
+  getCachedObservableQuery,
+  invalidateCachedObservableQuery,
+} from './queryCache';
+import {compact, objToKey, actHack} from './utils';
+
+export default function useQuery<Data = any, Variables = OperationVariables>(
+  queryOrComponent: DocumentNode<Data, Variables>,
+  options?: QueryHookOptions<Data, Variables>,
+): QueryHookResult<Data, Variables>;
+
+export default function useQuery<
+  Data = any,
+  Variables = OperationVariables,
+  DeepPartial = {}
+>(
+  queryOrComponent: AsyncQueryComponentType<Data, Variables, DeepPartial>,
+  options?: QueryHookOptions<Data, Variables>,
+): QueryDocumentResolving<Data, Variables> | QueryHookResult<Data, Variables>;
+
+export default function useQuery<
+  Data = any,
+  Variables = OperationVariables,
+  DeepPartial = {}
+>(
+  queryOrComponent:
+    | DocumentNode<Data, Variables>
+    | AsyncQueryComponentType<Data, Variables, DeepPartial>,
+  options: QueryHookOptions<Data, Variables> = {},
+): QueryHookResult<Data, Variables> | QueryDocumentResolving<Data, Variables> {
+  const {
+    skip = false,
+    pollInterval,
+    notifyOnNetworkStatusChange = false,
+    client: overrideClient,
+    context,
+    variables,
+    fetchPolicy,
+    errorPolicy,
+  } = options;
+
+  const [responseId, setResponseId] = useState(0);
+  const client = useApolloClient(overrideClient);
+  const query = useGraphQLDocument(queryOrComponent);
+
+  const watchQueryOptions: WatchQueryOptions<Variables> = useMemo(
+    () =>
+      compact({
+        context,
+        errorPolicy,
+        fetchPolicy,
+        notifyOnNetworkStatusChange,
+        pollInterval,
+        query,
+        variables,
+      }),
+    [
+      query,
+      pollInterval,
+      notifyOnNetworkStatusChange,
+      context && objToKey(context),
+      variables && objToKey(variables),
+      fetchPolicy,
+      errorPolicy,
+    ],
+  );
+
+  const observableQuery = useMemo(
+    () => {
+      if (!query) {
+        return;
+      }
+
+      // eslint-disable-next-line consistent-return
+      return getCachedObservableQuery<Data, Variables>(
+        client,
+        watchQueryOptions,
+      );
+    },
+    [query, client, watchQueryOptions],
+  );
+
+  useEffect(
+    () => {
+      if (skip || !observableQuery) {
+        return;
+      }
+
+      const invalidateCurrentResult = () => {
+        // A hack to get rid React warnings during tests. The default
+        // implementation of `actHack` just invokes the callback immediately.
+        // In tests, it's replaced with `act` from react-testing-library.
+        // A better solution welcome.
+        actHack(() => {
+          setResponseId(x => x + 1);
+        });
```

Comment:
Can you tell me more about what you mean here and why you are suggesting using `useRef` for setting response ID? Is it because we want to value to be shared and incremented across all instances of `useQuery`? Since it is a state, it will start from `0` for every instance and defeat the purpose of `setResponseId` being unique. Am I understanding this correctly?

Reply:
talked to Chris IRL to resolved this.

Basically, this is a mechanism to ensure if a few query with the same options are running back to back, the result that's being return is actually the one being called. 

Since currentResult is being memoize, it won't change unless the dependency change, which mean it won't get change if you call two exactly the same query twice (without responseId)

There are other ways to accomplish this without it being as confusing, (ie. store `currentResult` in state instead), I am still weighting the options.

Reply:
Not sure if this is relevant:
In the return of the `useEffect`, we unsubscribe. I wonder if it is possible to cancel the pending request (as part of this unsubscribe) if a particular dependency changes (maybe request id, maybe something else)

In any case, thanks for the context. :)